code/parachain/runtime/common/src/dex.rs

```diff
@@ -0,0 +1,262 @@
+use crate::{cosmwasm::*, prelude::*};
+use composable_traits::dex::*;
+use cosmwasm_std::{to_binary, Coin, QueryResponse, Response, Uint128};
+use primitives::currency::CurrencyId;
+use sp_runtime::traits::Convert;
+use sp_std::marker::PhantomData;
+
+pub struct DexPrecompile<Balance, AccountId, Dex>(PhantomData<(Balance, AccountId, Dex)>);
+
+pub enum CosmwasmPrecompileError {
+	AssetConversion,
+	AccountConversion,
+	DispatchError,
+	Serde,
+}
+
+impl<Dex, Balance, AccountId> DexPrecompile<Balance, AccountId, Dex>
+where
+	Dex:
+		Amm<AssetId = CurrencyId, Balance = Balance, AccountId = AccountId, PoolId = crate::PoolId>,
+	Balance: From<Uint128> + Into<u128>,
+	AccountId: sp_std::convert::From<sp_runtime::AccountId32>,
+{
```

Comment:
Wouldn’t that work:

```suggestion
pub struct DexPrecompile<Dex>(PhantomData<Dex>);

pub enum CosmwasmPrecompileError {
	AssetConversion,
	AccountConversion,
	DispatchError,
	Serde,
}

impl<Dex> DexPrecompile<Dex>
where
	Dex: Amm<AssetId = CurrencyId, PoolId = crate::PoolId>,
	Dex::Balance: From<Uint128> + Into<u128>,
	Dex::AccountId: sp_std::convert::From<sp_runtime::AccountId32>,
{
```
I don’t think having the `Balance` and `AccountId` generic arguments gives us anything since they can be deduced from `Dex`.